src/libostree/ostree-repo-finder-mount.c

```diff
@@ -256,122 +433,66 @@ ostree_repo_finder_mount_resolve_async (OstreeRepoFinder                  *finde
-              g_clear_error (&local_error);
-              continue;
-            }
-
-          /* Check the resolved repository path is below the mount point. Do not
-           * allow ref symlinks to point somewhere outside of the mounted
-           * volume. */
-          if (stbuf.st_dev != mount_root_stbuf.st_dev)
-            {
-              g_debug ("Ignoring ref (%s, %s) on mount ‘%s’ as it’s on a different file system from the mount.",
-                       refs[i]->collection_id, refs[i]->ref_name, mount_name);
-              g_clear_error (&local_error);
-              continue;
-            }
-
-          /* Exclude repositories which resolve to @parent_repo. */
-          if (stbuf.st_dev == parent_repo->device &&
-              stbuf.st_ino == parent_repo->inode)
+          for (gsize j = 0; j < repos_refs->len; j++)
             {
-              g_debug ("Ignoring ref (%s, %s) on mount ‘%s’ as it is the same as the one we are resolving",
-                       refs[i]->collection_id, refs[i]->ref_name, mount_name);
-              g_clear_error (&local_error);
-              continue;
+              const RepoAndRefs *repo_and_refs = &g_array_index (repos_refs, RepoAndRefs, j);
+              OstreeRepo *repo = repo_and_refs->repo;
+              GHashTable *repo_refs = repo_and_refs->refs;
+              g_autofree char *repo_path = g_file_get_path (ostree_repo_get_path (repo));
+
+              const gchar *checksum = g_hash_table_lookup (repo_refs, refs[i]);
+
+              if (checksum == NULL)
+                {
+                  g_debug ("Ignoring repository ‘%s’ when looking for ref (%s, %s) on mount ‘%s’ as it doesn’t contain the ref.",
+                           repo_path, refs[i]->collection_id, refs[i]->ref_name, mount_name);
+                  g_clear_error (&local_error);
+                  continue;
+                }
+
+              /* Finally, look up the GPG keyring for this ref. */
+              keyring = ostree_repo_resolve_keyring_for_collection (parent_repo, refs[i]->collection_id,
+                                                                    cancellable, &local_error);
+
+              if (keyring == NULL)
+                {
+                  g_debug ("Ignoring repository ‘%s’ when looking for ref (%s, %s) on mount ‘%s’ due to missing keyring: %s",
+                           repo_path, refs[i]->collection_id, refs[i]->ref_name, mount_name, local_error->message);
+                  g_clear_error (&local_error);
+                  continue;
+                }
+
+              /* There is a valid repo at (or pointed to by)
+               * $mount_root/.ostree/repos.d/$something.
+               * Add it to the results, keyed by the canonicalised repository URI
+               * to deduplicate the results. */
+              g_autofree char *canonical_repo_path = realpath (repo_path, NULL);
```

Comment:
Not new in this patch but I don't quite get why we need to canonicalize; feels like a leftover from the previous symlinks schema?

Reply:
There are still symlinks in this schema (although fewer of them); canonicalisation allows us to eliminate the duplicate arising from (for example):
 * `$usb_stick/ostree/repo` being a repository
 * `$usb_stick/.ostree/repos.d/alias` → `$usb_stick/ostree/repo` being a symlink
or the duplicate arising from two symlinks in `$usb_stick/.ostree/repos.d` pointing to the same repository elsewhere on the stick.

Is there any downside to canonicalising? I guess it’s not dirfd-relative, which is a a bit of a pain, but I don’t think `realpathat()` exists. I can write a libglnx version of `realpathat()` if you’re concerned about that.

Reply:
We could canonicalize by making a hashset - maybe add `ostree_repo_hash` too since you just added `_equals()`?  Anyways, doesn't have to be done in this patch.

Reply:
I’ve opened issue #1191 as a follow-up.